Add Vim9 generic function support (#17313)
vim/vim: runtime/doc/vim9.txt

2.  Differences				|vim9-differences|
3.  New style functions			|fast-functions|
4.  Types				|vim9-types|

> 5.  Generic functions			|generic-function|

Reviewer: ```suggestion
5.  Generic functions			|generic-functions|
```

Author: Thanks.  Updated the help text with this change.

==============================================================================

5. Generic functions

						*generic-function*
A generic function allows using the same function with different argument
types, while retaining type checking for arguments and the return value.  This
provides type safety and code reusability.

Declaration~
						*generic-function-declaration*
						*E1549* *E1550* *E1555*
The generic types for a generic function are declared in angle brackets "<"
and ">" directly after the function name. Multiple type names are separated by
commas.  These generic types can then be used like any other type within the

> function signature and body.  Example: >

Reviewer 2: ```suggestion
and ">" directly after the function name.  Multiple type names are separated
by commas.  These generic types can then be used like any other type within
the function signature and body.  Example: >
```

Author: Thanks. Updated the help text incorporating these comments.

    def MyFunc<T, A, B>(arg1: T): T
	var f: A
	var x = arg1
	return x
    enddef
<
						*generic-type-naming* *E1548*
The convention is to use a single uppercase letter for a generic type (e.g.,
T, A, X), although longer names are allowed. The name must start with an

> uppercase letter.

Reviewer 2: ```suggestion
T, A, X), although longer names are allowed.  The name must start with an
uppercase letter.
```

						*E1554* *E1556*
A function must be declared and used either as generic or as a regular
function - but not both.

						*E1557*
A generic type name must not conflict with other defined names, such as class
names, type aliases, enum names, function names or other generic type names.

Calling a generic function~
						*generic-function-call*
To call a generic function, specify the concrete types in "<" and ">"
between the function name and the argument list: >

    MyFunc<number, string, list<number>>()
<
						*E1551* *E1552* *E1553*
The number of concrete types provided when calling the function must match the

> number of generic types defined. An empty type list is not allowed.

Reviewer 2: ```suggestion
number of generic types defined.  An empty type list is not allowed.
```

Any Vim9 type (|vim9-types|) can be used as a concrete type in a generic
function.

Spaces are not allowed between the function name and "<", or between ">" and
the opening "(".

A generic function can be exported and imported like a regular function.
See |:export| and |:import|.

A generic function can be defined inside another regular or generic function.

Nested generic types~

Generic types can contain other generic types. This is useful for complex data

> structures like lists of dictionaries or dictionaries of lists. Example: >

Reviewer 2: ```suggestion
Generic types can contain other generic types.  This is useful for complex
data structures like lists of dictionaries or dictionaries of lists.
Example: >
```

    vim9script

    def Flatten<T>(x: list<list<T>>): list<T>
	var result: list<T> = []
	for inner in x
	    result += inner
	endfor
	return result
    enddef

    echo Flatten<number>([[1, 2], [3]])
<
Generic class method~

A Vim9 class method can be a generic function: >

    class A
	def Foo<A, B>()
	enddef
    endclass
    var a = A.new()
    a.Foo<number, string>()
<
Generic function reference~

> A function reference (|Funcref|) can be a generic function. This allows for

Reviewer 2: ```suggestion
A function reference (|Funcref|) can be a generic function.  This allows for
```

==============================================================================

						*generic-functions*
5. Generic functions

A generic function allows using the same function with different argument

> types, while retaining type checking for arguments and the return value.  This

Reviewer 3: ‘type arguments’ are probably better than ‘argument types’,  
because this is about passing different types at call sites  
to a function, whereas ‘argument types’ imply expressiveness  
of generic function declarations.

Author: Updated the help text.

provides type safety and code reusability.

Declaration~
						*generic-function-declaration*
						*E1550* *E1551* *E1556*
The generic types for a generic function are declared in angle brackets "<"
and ">" directly after the function name.  Multiple type names are separated
by commas: >

>     def[!] <{type} [, {types}]>{funcname}([arguments])[: {return-type}]

Reviewer 3: According to the above description first goes the name of  
a function, then goes the type parameter list:
```suggestion
    def[!] {funcname}<{type parameters}>([formal parameters])[: {return-type}]
```

Author: Updated the help text.

	{function body}
    enddef
<
These generic types can then be used like any other type within the function
signature and body.  Example: >

>     def MyFunc<T, A, B>(arg1: T): T

Reviewer 3: ```suggestion
    def MyFunc<T, A, B>(param1: T): T
```

Author: Updated the help text.

	var f: A

> 	var x = arg1

Reviewer 3: ```suggestion
	var x = param1
```

Author: Updated the help text.

A generic function can be defined inside another regular or generic function.

> Nested type variables~

Reviewer 3: ```suggestion
Referencing type variables in generic types~
```

Author: Updated the help text.

> Types variables can contain other type variables.  This is useful for complex

Reviewer 3: ```suggestion
Instead of concrete types, type variables can be used with generic types.
This is useful for complex data structures like lists of dictionaries or
dictionaries of lists.  Example: >
```

Author: Updated the help text.

Limitations and Future Work~

Currently, Vim does not support:
   - Type inference for type variables: All types must be explicitly specified
     when calling a generic function.
   - Type constraints: It's not possible to restrict a type variable to a
     specific class or interface (e.g., `T extends SomeInterface`).

>    - Default type variable arguments: Providing a default type for a generic

Reviewer 3: ```suggestion
   - Default type arguments: Providing a default type for a type parameter
     when not explicitly specified.
```

Author: Updated the help text.
